### ISSUE TITLE
xs:sequence with maxOccurs ?

### ISSUE THREAD

**Issue by cboettig** [cboettig/eml2#3]
@mbjones Quick question for you: 

When you see an `<xs:choice` element that has a `maxOccurs` attribute, is that setting the `maxOccurs` for all the child `<xs:element`s in the `<xs:choice`?  e.g. https://github.com/cboettig/eml2/blob/master/inst/xsd/eml-access.xsd#L262

(Until now I hadn't realized that `maxOccurs` could occur / be inherited from other levels instead of appearing on the `<xs:element` itself...

**Comment by mbjones** [comment]
@cboettig No, the cardinality on the containing `xs:sequence` or `xs:choice` tells you how many times you can repeat the choice -- setting maxOccurs > 1 on an `xs:choice` allows the children elements to repeat in any order. For example, the choice model of `A (B | C)*` means that A can contain any number of B or C elements in any order, whereas `A (B, C)*` represents a repeating sequence in which A can contain B then C, and repeat that sequence as needed.  These both totally differ from setting the cardinality on the child elements, for example the choice `A (B* | C)` allows A to contain any number of B elements or one C element. Not all models that can be expressed by the syntax are valid models, especially those that are not deterministic.

**Comment by cboettig** [comment]
Is it a show-stopper if we don't preserve element order among repeated elements?  It's pretty challenging to work with XML where a lot of data is stored only implicitly in the ordering of elements rather than explicitly... (e.g. the NeXML schema for evolutionary data insists that all data is explicit rather than implicit in ordering).  It's not always clear to me when a different ordering does or does not actually convey anything semantically different.  

On the plus side, the tests you provide are very helpful in tweaking my schema hacking generation of classes.  I believe we can now parse EML files in eml2 that we couldn't parse in EML, so that's progress.

**Comment by mbjones** [comment]
I'd be very careful about changing order.  The XSD validators are really sensitive to that, and I'll bet there are places where it will trip up the validator if you change order of elements.  Beyond validation, I can think of two issues: 1) there is semantic meaning to the order of `<attribute>` elements inside of an `<attributeList>` element -- the order is supposed to conform to the order in which those attributes would occur in the physically formatted file, such as a CSV file, and shouldn't be changed; and 2) there is semantic meaning to the order of `<allow>` and `<deny>` elements in access control lists.  

I think you should be able to do a full round trip through the parse/serialze and not change the order of anything or lose any content.  One way to check that would be to use an XML parser to convert both your input and output to [XML canonical form](http://www.xml.com/pub/a/ws/2002/09/18/c14n.html) and see if those documents are identical.

**Comment by cboettig** [comment]
Right, roundtrip would be a good goal.  And thanks for the tip about 'canonical form', I was wondering how to do the identity comparison.  

I am afraid this is going to be pretty difficult in R.  For instance, the current approach of mapping child elements into slots in an S4 object doesn't preserve the order.  e.g. if you define an S4 class `A` with slots `B` and `C`, you get the same class from parsing 

```xml
<A>
  <B>stuff</B>
  <C>more</C>
</A>
```

as you would from:

```xml
<A> 
  <C>more</C>
  <B>stuff</B>
</A>
```

even though we would preserve the order of any repeated element, since they map to a list.  So the roundtrip proposal seems a bit frustrating, because it's not clear that there's ever a case where the order of non-identical child elements conveys metadata, and yet we might need an entirely new approach to preserve that non-semantic information, right?

**Comment by cboettig** [comment]
Still wrapping my head around this one: 

> For example, the choice model of A (B | C)* means that A can contain any number of B or C elements in any order,

So in `ResponsibleParty`, does that mean that all of these are valid?   


```xml
   <party>
      <individualName>
      <organizationName>
      <postitionName>
      <individualName>
      <organizationName>
      <postitionName>
```

```xml
   <party>
      <individualName>
      <individualName>
      <organizationName>
      <organizationName>
      <postitionName>
      <postitionName>
```


```xml
   <party>
      <individualName>
      <organizationName>
      <organizationName>
      <postitionName>
```

Do the first two then have the same meaning as well?  If so, it seems these are best represented by just allowing for repeating elements.  This would always generate a predictable order, (e.g. like case 2) rather than the order of the original file; but it's not obvious why such behaviour would be undesirable.  

I certainly do see your point that this is different for any `//xs:sequence[@maxOccurs]` elements  (looks like a total of 4 such elements exist in EML schema files: `citetitle`,  `role`, `methodStep`, `descriptorValue`), where order is semantic.  For these we probably need a shadow list type to preserve ordering of the suite of elements without having to wrap them in a parent element.

**Comment by cboettig** [comment]
So I think the strategy here is just to make sure we get the slot names for elements coming from any `xs:sequence` need to be given in order of that sequence. In the case of `choice` the s4 object has slots for both options [^1] and the order doesn't matter.  

This is mostly done; though I need to test more thoroughly that all classes declare the elements in the correct sequence order.  (Notably, this means we cannot inherit classes from `contains` since these are always placed after the slots defined by `slots`.  To address this, the S4 object adds a 'metaslot' for any xs:group to ensure those elements are grouped and appear in the correct order.  

It remains to implement the shadow lists for the 4 cases of `//xs:sequence[@maxOccurs]`

[^1] : and actually doesn't enforce that only one of those is occupied -- we could add this in along with loads of other validation checks to the class's validator method, but for now will leave validation to be done by xmlSchemaValidate instead.

**Comment by mbjones** [comment]
@cboettig Better late than never... but I think you need to preserve the order in cases like the party repeating choice.  Although `xs:choice` allows any order, the order for elements that authros choose can still have meaning and should never be lost.  People will often enter fields in the order that is most significant to them, and may include, for example, their name, department name, and university name, and they wouldn't wan't those fields to get rearranged into a different order.  In general, the eml package needs to 1) preserve order of elements that users have set, and 2) allow users to reorder elements as needed (e.g., change the order of creators in a dataset to indicate citation precedence).  I'm going to reopen this issue because I think the order of choice elements still needs to be addressed.

**Comment by cboettig** [comment]
Okay, will look into this.  We can capture the order when parsing and preserve that in a dedicated slot which we obey when serializing again.

It's less clear to me how we handle this for the constructor functions, or if the user modifies the object by adding or removing slots (particularly if they are manipulating the S4 object directly).  

Anyway, will try and come up with something.  I guess since the earlier `EML` package didn't handle this either; we need not include this issue in the milestone to replace `EML` with `eml2` ;)
